Initial take on a potential activity.mojom replacement

chinmaygarde opened the pull request.

abarth commented on sky/services/platform/filesystem.mojom:
```diff
@@ -0,0 +1,38 @@
+// found in the LICENSE file.
+
+[DartPackage="sky_services"]
+module filesystem;
```
We should give these more esoteric module names.  I would have expected a filesystem module to be an actual file system (e.g., read/write/flush).

module flutter.platform

?

abarth commented on sky/services/platform/filesystem.mojom:
```diff
@@ -0,0 +1,38 @@
+module filesystem;
+
+/// Returns commonly used locations on the filesystem
+interface Filesystem {
```
This isn't really a file system.  It's a way to get some well-known paths.  Maybe `PathProvider` ?

krisgiesing replied:
Or DirectoryProvider, since all the methods are named *Directory().

abarth commented on sky/services/platform/interface.mojom:
```diff
@@ -0,0 +1,95 @@
+// found in the LICENSE file.
+
+[DartPackage="sky_services"]
+module platform_interface;
```
interface is redundant.  Everything in mojom is an interface.  I'd go with `module flutter.platform` in all these mojoms.

abarth replied:
Maybe `flutter.environment`

abarth commented on sky/services/platform/interface.mojom:
```diff
@@ -0,0 +1,95 @@
+};
+
+/// Controls specific aspects of the embedder interface
+interface Interface {
```
`s/Interface/Chrome/`

Chrome is a term that refers to the decorations around the main content.

krisgiesing replied:
Chrome is a little overloaded.  Maybe SystemChrome?

chinmaygarde replied:
Done

Hixie commented on sky/services/platform/interface.mojom:
```diff
@@ -0,0 +1,95 @@
+/// A mask of device interface orientations is created by or'ing individual
+/// options
+enum /* options*/ DeviceOrientation {
+  PortraitUp      = 1,
```
(some systems, e.g. phones, won't let you do PortraitDown)

chinmaygarde replied:
Maps to `UIDeviceOrientationPortraitUpsideDown` on iOS. If the platform cannot support the orientation or if the combination of options cannot be statisfied (say on Android in this case), the SetPreferredOrientations method would return false.

chinmaygarde commented:
I have updated the names per the suggestions

Hixie commented on sky/services/platform/filesystem.mojom:
```diff
@@ -0,0 +1,38 @@
+[DartPackage="sky_services"]
+module filesystem;
+
+/// Returns commonly used locations on the filesystem
```
always end every sentence in the docs with a period. (applies throughout)

chinmaygarde replied:
Done

Hixie commented on sky/services/platform/interface.mojom:
```diff
@@ -0,0 +1,95 @@
+/// Specifies a particular device orientation
+/// A mask of device interface orientations is created by or'ing individual
+/// options
+enum /* options*/ DeviceOrientation {
```
I would order these as PortraitUp, LandscapeLeft, PortraitDown, LandscapeRight.
I am not sure what the names should be.
Maybe document it as:
```
/// To determine which values correspond to which orientations, first position the device
/// in its default orientation (this is the orientation that the system first uses for its
/// boot logo, or the orientation in which the hardware logos or markings are upright, or
/// the orientation in which the cameras are at the top). If this is a portrait orientation,
/// then this is PortraitUp. Otherwise, it's LandscapeLeft.
/// As you rotate the device by 90 degrees in a counter-clockwise direction around the
/// axis that traverses the screen, you step through each value in this enum in the order
/// given. (For a device with a landscape default orientation, the orientation obtained by
/// rotating the device 90 degrees clockwise from its default orientation is PortraitUp.)
```

chinmaygarde replied:
Good idea, its best to ambiguity w.r.t this thing. Especially when dealing with all the different variants on phones and tablets

Hixie commented on sky/services/platform/interface.mojom:
```diff
@@ -0,0 +1,95 @@
+  LandscapeRight  = 8,
+};
+
+/// Specifies how the status bar(s) should be displayed
```
This is very hard to be generic about. Android, Chrome OS, Windows, Mac OS, iOS, they're all different. Every window manager on X11 has a different idea too. But maybe:

```
enum SystemUIOverlays {
  Top      = 1,
  Bottom   = 2,
  Floating = 4,
  All      = 0xFFFFFFFFFFFFFFFF
}
```

...and you or them together to say which should be hidden.

chinmaygarde replied:
What does `Floating` map to?

Hixie replied:
like, the Facebook chat bubbles. If we can't disable those then skip it. I
was mostly just trying to show it could be extended.

On Thu, Jan 28, 2016, 17:05 Chinmay Garde <notifications@github.com> wrote:

> In sky/services/platform/interface.mojom
> <https://github.com/flutter/engine/pull/2315#discussion_r51213574>:
>
> > +enum /* options */ InterfaceOrientation {
> > +  Portrait  = 1,
> > +  Landscape = 2,
> > +};
> > +
> > +/// Specifies a particular device orientation
> > +/// A mask of device interface orientations is created by or'ing individual
> > +/// options
> > +enum /* options*/ DeviceOrientation {
> > +  PortraitUp      = 1,
> > +  PortraitDown    = 2,
> > +  LandscapeLeft   = 4,
> > +  LandscapeRight  = 8,
> > +};
> > +
> > +/// Specifies how the status bar(s) should be displayed
>
> What does Floating map to?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/flutter/engine/pull/2315/files#r51213574>.
>

Hixie commented on sky/services/platform/interface.mojom:
```diff
@@ -0,0 +1,95 @@
+  /// Return Value:
+  ///   boolean indicating if the preference was conveyed successfully to the
+  ///   embedder.
+  SetStatusBarStyle(StatusBarsStyle style) => (bool success);
```
I'd have this be a method that says what parts of the system UI you want to hide, using the mask I described above.

Hixie commented on sky/services/platform/filesystem.mojom:
```diff
@@ -0,0 +1,38 @@
+  ///
+  /// Example(s):
+  ///    iOS: The application bundle directory
+  AssetsDirectory() => (string path /* non-null */);
```
why `/* non-null */` ? I thought you had to be explicit about nullability in mojom.

chinmaygarde replied:
I guess I wanted to convey that the platform will always provide these directory locations. But this is confusing. Removed.

chinmaygarde commented:
Have addressed all feedback:
* Clarified behavior when the application requests showing or hiding unsupported overlays (on the platform).
* For enums specifying options, 0 is always an alias for all options so there is no separate item with ~0.
* Misc. doc + naming updates.

Another look please.

krisgiesing commented on sky/services/platform/system_chrome.mojom:
```diff
@@ -0,0 +1,113 @@
+  ///
+  /// Arguments:
+  ///   style: A mask of `SystemUIOverlay` enum values that denotes the overlays
+  ///          to show. A value of 0 is synonymous with all overlays enabled.
```
If the mask indicates which overlays to show, then shouldn't 0 be all overlays disabled?

chinmaygarde replied:
Derp. You're right. Will get rid of that wording.

krisgiesing commented on sky/services/platform/path_provider.mojom:
```diff
@@ -0,0 +1,38 @@
+  /// directory. Instead, the caller is responsible for creating
+  /// (and cleaning up) files or directories within this directory. It is
+  /// recommended that applications create subdirectories within this directory
+  /// prefixed by their application identifiers and place their temporary files
```
Aren't the temporary directories already sandboxed to the application? We might want to enforce that this always returns a sandboxed directory, even for implementations (e.g. desktop) where that's not inherently done.

chinmaygarde replied:
I am not sure its necessary for us to enforce that guarantee. That guarantee may be difficult to enforce.

krisgiesing commented on sky/services/platform/path_provider.mojom:
```diff
@@ -0,0 +1,38 @@
+  ///
+  /// Example(s):
+  ///    iOS: The application bundle directory.
+  AssetsDirectory() => (string path);
```
I'm not entirely sure we want to expose this as a path. For some abstractions, this content isn't available in a literal directory. Is there client code that can't use AssetBundle?

abarth replied:
Yeah, we shouldn't expose this.  We want to be able to serve assets out of the zip file or transcode to some other format.

chinmaygarde replied:
Good point, I thought this could be `getFilesDir` on Android but that is not actually the assets directory. But this is quite obviously wrong. If everything uses the asset bundle, can get rid of this.

krisgiesing commented:
Thanks for tackling this, looking much better than activity.mojom :)

chinmaygarde commented:
More amends:
* 0 is all disabled on system overlays.
* Temporary directories are scoped per application
* No "AssetsDirectory" accessor

krisgiesing commented:
LGTM

chinmaygarde commented:
Rebase. Thanks.

j9brown commented:
I only skimmed this but I have the feeling it's still all very Android-y.
It might make sense to split this layer into two pieces, a platform interop
layer which actually deals in native platform concepts, and a set of
framework abstractions which map some useful subset of those things into
higher level policies.

This way we don't end up trying to express the union of all idiosyncrasies
in one place.

Keep in mind that the structure of any given API is often informed as much
by who worked on what (historically) as any other internal logic it might
have been imagined to possess (and perhaps lost along the way).

On Fri, Jan 29, 2016, 4:53 PM Chinmay Garde <notifications@github.com>
wrote:

> Merged #2315 <https://github.com/flutter/engine/pull/2315>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/flutter/engine/pull/2315#event-532137453>.
>

chinmaygarde commented:
> I have the feeling it's still all very Android-y

Thats an interesting (and unexpected) observation. I took the `UIKit` concept of `UIInterfaceOrientationMask`s on `UIViewController`s and their interaction with `UIDeviceOrientation` and worked my way back to ensure that what we settled on also made sense on Android. So my intuition was exactly the opposite of yours. Same goes for the path provider. I just happen to be more familiar with `UIKit` concepts for now.

Per my reckoning, these `mojom`s express fairly high level abstractions and are platform API agnostic. The service provider implementations on each platform will have to deal with actual native platform concepts. So we will already have two distinct layers. Are you suggesting that we have not made the distinction clear(er)? If so, I can work on fixing that with some guidance.

Hixie commented:
@j9brown Can you elaborate on what specific parts are too Android-specific?
